src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double GROW_FACTOR = 1.5;
+    private static final int NOT_FOUND = -1;
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            elements = (T[]) new Object[initialCapacity];
+        } else {
+            throw new IllegalArgumentException("Illegal initial capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        growIfNeed();
 
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        isIndexValid(index, index > size);
```

Comment:
The only place where condition differs is here. Put that exclusive check here, without delegating to utility checkers, and then you can remove that weird second argument :) If you had more "diversity" in terms of acceptable limits, it would make sense to pass said limit (if the condition is >= you would pass size + 1 here and size everywhere else ) - passing a boolean is a bit unusual and redundant - you pass index twice